### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by vasco-santos**
This module provides the `go-libp2p-daemon` binary according to the platform where it is fetched (linux, mac os and windows).

### CODE REVIEW

**Comment by jacobheun** [comment]
I'm wondering if we should just make this package more generic, so we could pull other packages as need. It would be nice to have a single package that could pull go-ipfs, go-libp2p, go-daemon, etc. Looking through the code it doesn't seem like it would be much more effort to have the parent package.json file list the appropriate contents there (version, cid, etc). Might be something to table for later though, as it doesn't need to block this.

**Comment by vasco-santos** [comment]
@jacobheun 

I understand that your suggestion would be useful. However, I see some blockers with the needs of this solution. 

Firstly, the way to track versions of the dependencies. For instance, I started this repo with `6.0.30`, since it is the current version of `go-libp2p`. This way, when a new `go-libp2p` release happens, we will need to update this module to use the new version of `go-libp2p`, and use the same module version. So, I cannot think about other easy solution to track the version of multiple dependencies in a single module.

Other than that, in this current implementation, it is necessary to add three binary files to ipfs each time `go-libp2p` gets a release, being this necessary for interop tests. That is one binary file per platform (os x, linux and windows). Having several other packages here would need a higher maintenance cost.

We can discuss this later on, as I agree that it would be useful to have a more generic package with the latest versions of each module.

**Comment by jacobheun** [comment]
So, my thought was to avoid tracking the versions of the go libraries in here, and instead pull those from the package.json of the application, libp2p/interop, in this case. You wouldn't need to keep updating here and then pulling the version in there. We could just have in `libp2p/interop` package.json something like,
```json
"goDependencies": {
  "go-libp2p": "QmSS6iQ3JNJ96g73tFARAcdYaMmAsvjX1RVuMPqJThTs7D"
}
```
and then read the version from there when doing the download. Anytime we need a new version, we just add it to the package.json of the repo that actually wants it. This could also be a multi version mapping, in case we wanted to be able to fetch multiple versions.
```json
"goDependencies": {
  "go-libp2p": {
   "v6.0.30": "QmSS6iQ3JNJ96g73tFARAcdYaMmAsvjX1RVuMPqJThTs7D"
  }
}
```

**Comment by vasco-santos** [comment]
My biggest concern is having to provide binaries for each OS, such as:

```js
module.exports = {
  'v6.0.30': {
    'darwin': 'QmSS6iQ3JNJ96g73tFARAcdYaMmAsvjX1RVuMPqJThTs7D',
    'linux': 'Qmdu8WCFUW43u8mHpbj2jV2ZSB1mP7JcjnJJcJ1oF6iMwW',
    'win32': 'QmP5P6AYQjbG26j5qrzbohSqAW3AP2Mvsuun3fS99hC1x6'
  }
}
```

Anyway, we can probably discuss this in a issue and move in that way, having a fallback here in case the hashes are not provided.

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    const isWindows = platform === 'win32'
+
+    // Create the download url
+    const fileExtension = isWindows ? '.zip' : '.tar.gz'
+    const fileName = 'libp2p_' + version + fileExtension
+    const url = `${distUrl}/ipfs/${libp2pVersions[version][platform]}`
+
+    // Success callback wrapper
+    // go-ipfs contents are in 'go-ipfs/', so append that to the path
```
Comment (jacobheun) [inline src/index.js]:
```suggestion
    // go-libp2p contents are in 'go-libp2p/', so append that to the path
```

**Review by jacobheun — APPROVED** [review]
1 minor comment correction, but otherwise this looks good.

**Pull request merged**